File: 卡码/代码随想录/Python/单调栈/LeetCode_0496.py
#496. 下一个更大元素 I
from typing import *

class Solution:
	def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
		res:List[int]=[-1]*len(nums1)
		st:List[int]=[]
		m:dict[int,int]=dict()
		for i in range(len(nums1)):
			m[nums1[i]]=i
		for i in range(len(nums2)):
			while len(st) and nums2[i]>nums2[st[-1]]:
				if nums2[st[-1]] in m:
					res[m[nums2[st[-1]]]]=nums2[i]
				st.pop()
			st.append(i)
		return res

File: 卡码/代码随想录/C++/单调栈/monotonicStack/LeetCode_0496.cpp
//496. 下一个更大元素 I
#include <iostream>
#include <stack>
#include <unordered_map>
#include <vector>
using namespace std;

class Solution {
public:
	vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
		stack<int> st;
		unordered_map<int,int> m;
		vector<int> res(nums1.size(),-1);
		for(int i=0;i<nums1.size();i++) m[nums1[i]]=i;
		for(int i=0;i<nums2.size();i++){
			while(!st.empty()&&nums2[i]>nums2[st.top()]){
				if(m.find(nums2[st.top()])!=m.end()) res[m[nums2[st.top()]]]=nums2[i];
				st.pop();
			}
			st.push(i);
		}
		return res;
	}
};

int main(){
  system("pause");
  return 0;
}
